hubstorage/client.py

```diff
@@ -21,22 +42,48 @@ class HubstorageClient(object):
     DEFAULT_TIMEOUT = 60.0
 
     def __init__(self, auth=None, endpoint=None, connection_timeout=None,
-            max_retries=0):
+            max_retries=3):
         self.auth = xauth(auth)
         self.endpoint = endpoint or self.DEFAULT_ENDPOINT
         self.connection_timeout = connection_timeout or self.DEFAULT_TIMEOUT
-        self.session = self._create_session(max_retries)
+        self.session = self._create_session()
+        self.retrier = Retrying(stop_max_attempt_number=max_retries + 1, retry_on_exception=_hc_retry_on_exception)
```

Comment:
The idea is to expose only max_retries and max_delay to users and tune exponential multiplier to make them compatible. Considering `sum(2 ** x for x in range(max_retries)) == 2 ** max_retries - 1` the "compatible" multiplier can be computed  as `max_delay / (2 ** max_retries - 1)`.

> AFAIK the library is used for long running jobs so no one wants to fail sooner with an error rather than trying to finish the job. (the current DEFAULT_TIMEOUT isn't used).

The library is used by interactive applications too, like Dash and Deck, long wait times are not desirable in such cases but a few retries (3) in a short time (60s) is still responsive and useful to filter sporadic errors. In fact, when users access collections trough Dash UI what interacts with HS is this library.

 Long running jobs are a valid case where we can use a greater max_delay, this is why it should be configurable IMO, defaults to interactive case but let users configure it for when time is not a problem. 

> What is a useful value for max_delay?

I prefer one that favours the interactive case, like current 60s timeout. 

> max_retries * 50s (haproxy timeout) + exponential backoff? so more than 150s in the default case?

No.

> the current default? 60s, so that queries that timeout (the reason for this patch) are retried only once?

Yes, unless we start talking about setting delays and timeouts per call. As explained before, when used in jobs, the user can increase the delay to his desire. 

> if the user sets only max_retry to a new value, should we change the max_delay to accommodate it?

No, we set exponential multiplier accordingly.
 
> if the user sets both max_retry and max_delay, should we warn when they are incompatible?

NA.

> if the user sets max_delay to a small value like 10s should we warn?

No. But max_delay is a bad name, It should be clear the delay is for retries. i.e.: max_retry_time


> Setting an exponential multiplier and a jitter for the sake of hubstorage/jobq is a good idea, 

I agree.

> but there's no reason to expose these 2 options to the user, we own the API.

My reasons are explained above, not imposed though.
 
> Something like 500ms for the exponential backoff? (wait 1, 2 then 4 seconds between requests).
> And a jitter to 500ms? sound's good?

Sounds good, although in other components we use a jitter that is proportional to the wait time between retries. I don't think it is possible with retrying lib.